web/api/web_api_v1.c

```diff
@@ -615,12 +615,67 @@ inline int web_client_api_request_v1_registry(RRDHOST *host, struct web_client *
     }
 }
 
+inline void web_client_api_request_v1_info_summary_alarm_statuses(const RRDHOST *host, BUFFER *wb) {
+    int alarm_normal = 0, alarm_warn = 0, alarm_crit = 0;
+    RRDCALC *rc;
+    for(rc = host->alarms; rc ; rc = rc->next) {
+        if(unlikely(!rc->rrdset || !rc->rrdset->last_collected_time.tv_sec))
+            continue;
+
+        switch(rc->status) {
+            case RRDCALC_STATUS_WARNING:
+                alarm_warn++;
+                break;
+            case RRDCALC_STATUS_CRITICAL:
+                alarm_crit++;
+                break;
+            default:
+                alarm_normal++;
+        }
+    }
+    buffer_sprintf(wb, "\t\t\"normal\": %d,\n", alarm_normal);
+    buffer_sprintf(wb, "\t\t\"warning\": %d,\n", alarm_warn);
+    buffer_sprintf(wb, "\t\t\"critical\": %d\n", alarm_crit);
+}
+
+inline void web_client_api_request_v1_info_mirrored_hosts(const RRDHOST *host, BUFFER *wb) {
+    RRDHOST *rc;
+    for(rc = host->next; rc ; rc = rc->next) {
```

Comment:
For this to be thread safe, you need to `rrd_rdlock()`.

Reply:
There is also a macro for that `rrdhost_foreach_read(rc)`, that also checks that a read lock really exists.

Reply:
Thank you. I've added it.